DOCSP-30552: Collations guide (#82)
mongodb/docs-rust: source/fundamentals/collations.txt

  numerical order rather than alphabetical order
- Specifies a collation in a ``FindOptions`` instance, which overrides the collection's default collation
- Uses the ``find()`` method to return documents in which the value of the ``length`` field is greater
  than ``"1000"``

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/collation.rs
      :language: rust
      :dedent:
      :start-after: start-op-collation
      :end-before: end-op-collation

   .. output::
      :language: none
      :visible: false

      { "name" : "Les Misérables", "length": "1462" }
      { "name" : "Infinite Jest", "length" : "1104" }
      { "name" : "Ça", "length" : "1138" }
   
If you run the preceding find operation without setting the ``numeric_ordering`` option to ``true``,
the driver compares ``length`` values as strings. Then, the code returns the following documents:

.. code-block:: json
   :copyable: false
   
   { "name" : "Emma", "length" : "474" }
   { "name" : "Les Misérables", "length": "1462" }
   { "name" : "Infinite Jest", "length" : "1104" }
   { "name" : "Cryptonomicon", "length" : "918" }

>    { "name" : "Ça", "length" : "1138" }

Reviewer: S: Instead of including the output, can you just give more detail in the preceding paragrah about what the matches are when they are compared as strings? For example, When the length values are compared as strings, 474 is considered alphabetically after 1000.

.. _rust-collations:

==========
Collations
==========

> 

Reviewer: S: add meta tags!

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use **collations** to order your find or
aggregation operation results by string values. A collation is a set of character

> ordering conventions that apply to a specific language and locale.

Reviewer: S: "that apply to" could be confusing here. Could you use "based on" or "that correspond to"?
The conventions actually apply to your data/results, so I think this sentence should be more clear.

This guide includes the following sections:

> - :ref:`rust-mongodb-collations` describes how MongoDB sorts strings according
>   to both the default collation and a custom collation

Reviewer: ```suggestion
- :ref:`rust-mongodb-collations` describes how MongoDB sorts string values according
  to the default collation and custom collations
```

> - :ref:`rust-specify-collation` describes how to create a ``Collation`` instance by
>   using builder functions 

Reviewer: S: remove some detail
```suggestion
- :ref:`rust-specify-collation` describes how to create a ``Collation`` struct instance
```

- :ref:`rust-collection-collation` describes how to set the collation for a
  new collection
- :ref:`rust-index-collation` describes how to set the collation for an index

> - :ref:`rust-op-collation` describes how to apply a collation to read, update, and
>   delete operations

Reviewer: ```suggestion
- :ref:`rust-op-collation` describes how to apply a collation to CRUD operations
```

- :ref:`rust-collation-addtl-info` provides links to resources and API documentation
  for types and methods mentioned in this guide

.. _rust-mongodb-collations:

MongoDB Collations
------------------

MongoDB sorts strings using *binary collation* by default. This collation

> method uses the `ASCII standard <https://en.wikipedia.org/wiki/ASCII>`_

Reviewer: S: use a permanent link. Also, per the style guide, make link text clear. Perhaps you can move the wiki to a tip
```suggestion
method uses the `ASCII standard <https://en.wikipedia.org/wiki/ASCII>`_
```

character values to compare and order strings. Certain languages and locales
have specific character ordering conventions that differ from the ASCII
standard.

For example, in Canadian French, the right-most accented character determines
the ordering for strings when the other characters are the same. Consider the
following Canadian French words: 

- cote
- coté
- côte
- côté

When using the default binary collation, MongoDB sorts the words in the following order:

.. code-block:: none

   cote
   coté
   côte
   côté

In this sort order, "coté" is placed before "côte" because the ASCII standard "o" character value

> is lower than the "ô" character value.

Reviewer: Q: what do you mean by "lower" here?

Reviewer: Should this be reversed?

Author: I meant a lower ASCII code value, but yeah it's confusing so I will reword

When using the Canadian French collation, MongoDB sorts the words in the following order:

.. code-block:: none

   cote
   côte
   coté
   côté

In this sort order, "coté" is placed after "côte" because Canadian French collation rules position
the character "e" before the character "é".

.. _rust-specify-collation:

Specify a Collation
-------------------

> You can define a collation by specifying your collation locale and related options in a

Reviewer: ```suggestion
You can define a collation by specifying a collation locale and other options in a
```

``Collation`` struct instance. To begin building a ``Collation`` instance, call the
``Collation::builder()`` method.

.. note:: Instantiating Options
   
   The {+driver-short+} implements the Builder design pattern for the
   creation of many different types, including ``Collation``. You can
   use the ``builder()`` method to construct an instance of each type

>    by chaining option builder functions one at a time.

Reviewer: ```suggestion
   by chaining option builder methods.
```

Reviewer: use method for consistency with docs set

> Call the following ``Collation::builder()`` functions to specify the collation. You must use
> the ``locale()`` function to build the ``Collation`` instance, but all other builder functions
> are optional:

Reviewer: ```suggestion
The following table describes the builder methods that you can use to set fields of a ``Collation`` instance.
You must use the ``locale()`` function to build a valid ``Collation`` struct, but all other builder functions
are optional:
```

       ``CollationAlternate::Shifted``
     - Specifies whether the driver considers whitespace and punctuation as base characters
       during string comparison

   * - | ``max_variable()``
     - | ``CollationMaxVariable::Punct``,
       | ``CollationMaxVariable::Space``
     - | Specifies which characters the driver ignores when ``alternate`` is set to
       | ``CollationAlternate::Shifted``

   * - ``normalization()``
     - ``true``, ``false``
     - Specifies whether the driver performs text normalization for string values

   * - | ``backwards()``
     - | ``true``, ``false``
     - | Specifies whether the driver sorts strings containing diacritics in reverse character order

Example
~~~~~~~

The following example specifies a ``Collation`` instance and sets the collation locale to ``"en_US"``:

.. literalinclude:: /includes/fundamentals/code-snippets/collation.rs
   :language: rust
   :dedent:
   :start-after: start-collation
   :end-before: end-collation

.. _rust-collection-collation:

Set a Collation on a Collection
-------------------------------

When you create a new collection, you can define the collation for future operations

> called on that collection. Set the collation by using the ``collation()`` function
> to build a ``CreateCollectionOptions`` instance. Then, call the ``create_collection()``

Reviewer: ```suggestion
called on that collection. Set the collation by using the ``collation()`` function
when creating a ``CreateCollectionOptions`` instance. Then, call the ``create_collection()``
```

method with your options instance as a parameter.

.. _rust-create-collection:

Create Collection with a Collation Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This example creates a collection called ``books`` and specifies a collation according

> to the ``"fr"``, or French, locale conventions. The ``strength`` field is set to

Reviewer: S: reverse the two clauses of this setnece to alogn with the order of the code

``CollationStrength::Primary`` to ignore differences in diacritics.

.. literalinclude:: /includes/fundamentals/code-snippets/collation.rs
   :language: rust
   :dedent:
   :start-after: start-create-collection
   :end-before: end-create-collection

> Use the Collection's Collation Example

Reviewer: S: right now, this title is a little hard to follow. Perhaps something like "Collation ordering demonstration" would be more apt? Happy to brainstorm!

Author: I was a little hesitant about just titling it "Collation Ordering Demonstration" since that doesn't mention that it's using the collation set for the collection, but since it's a subheading under "Set a Collation on a Collection" I feel like it does work -- changed!

.. code-block:: json

   { "name" : "Emma", "length" : "474" }
   { "name" : "Les Misérables", "length": "1462" }
   { "name" : "Infinite Jest", "length" : "1104" }
   { "name" : "Cryptonomicon", "length" : "918" }
   { "name" : "Ça", "length" : "1138" }

.. tip::

   To learn how to insert documents into a collection, see the :ref:`rust-insert-guide`
   guide.
   
The following example uses the ``find()`` method to return all documents in which the value
of the ``name`` field alphabetically precedes ``"Infinite Jest"``:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/collation.rs
      :language: rust
      :dedent:
      :start-after: start-default-query
      :end-before: end-default-query

   .. output::
      :language: none
      :visible: false

      { "name": "Emma", "length": 474 }
      { "name": "Cryptonomicon", "length": 918 }
      { "name" : "Ça", "length" : "1138" }
      

If you don't specify a collation for the ``books`` collection, the ``find()`` method follows
default binary collation rules to determine the ``name`` values that precede ``"Infinite Jest"``.

> These rules place words beginning with "Ç" after those beginning with "I". If the preceding find
> operation follows binary collation rules, the document that represents the book ``"Ça"`` is excluded
> from the results.

Reviewer: ```suggestion
These rules place words beginning with "Ç" after those beginning with "I". So, when the preceding find
operation follows binary collation rules, the document with the ``name`` value of ``"Ça"`` does not match
the filter criteria.
```

.. _rust-index-collation:

Set a Collation on an Index 
---------------------------

> When you create a new index on a collection, you can define the collation for operations
> that use the index. To run an operation that uses the index and its collation, your operation
> and index must specify the same collation. Additionally, ensure that the index specifying
> the collation covers the query. 

Reviewer: S: omit the last sentence by integrating it with the first sentence
```suggestion
When you create a new index on a collection, you can define the collation for operations
that are covered by the index. To run an operation that uses the index and its collation, your operation
and index must specify the same collation.
```

.. tip::

   To learn more about indexes and covered queries, see the :ref:`rust-indexes` guide.

Set the index collation by using the ``collation()`` function to build an ``IndexOptions`` instance.
Then, pass your ``IndexOptions`` as an argument to an ``IndexModel`` builder function, and pass your
``IndexModel`` as an argument to the ``create_index()`` method.

Example
~~~~~~~

The following example uses the ``create_index()`` method to create an ascending index on the

> ``name`` field and specifies a new collation with an ``"en_US"`` locale:

Reviewer: S: avoid with
```suggestion
``name`` field and specifies a new collation corresponding to the ``"en_US"`` locale:
```

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/collation.rs
      :language: rust
      :dedent:
      :start-after: start-index
      :end-before: end-index

   .. output::
      :language: none
      :visible: false

       Created index: name_1

.. _rust-op-collation:

Set a Collation on an Operation
-------------------------------

Operations that read, update, and delete documents from a collection can use collations.
Applying a collation to an operation overrides any collation previously defined for a

> collection.

Reviewer: ```suggestion
collection or index.
```

aggregation operation results by string values. A collation is a set of character
ordering conventions that correspond to a specific language and locale.

This guide includes the following sections:

- :ref:`rust-mongodb-collations` describes how MongoDB sorts string values according
  to the default collation and custom collations
- :ref:`rust-specify-collation` describes how to create a ``Collation`` struct instance
- :ref:`rust-collection-collation` describes how to set the collation for a
  new collection
- :ref:`rust-index-collation` describes how to set the collation for an index
- :ref:`rust-op-collation` describes how to apply a collation to certain CRUD operations
- :ref:`rust-collation-addtl-info` provides links to resources and API documentation
  for types and methods mentioned in this guide

.. _rust-mongodb-collations:

MongoDB Collations
------------------

MongoDB sorts strings using *binary collation* by default. This collation method
uses the ASCII standard character values to compare and order strings. Certain
languages and locales have specific character ordering conventions that differ from
the ASCII standard.

.. tip::

>    To learn more about the ASCII standard, see the `ASCII
>    <https://en.wikipedia.org/w/index.php?title=ASCII&oldid=1180396712>`_ Wikipedia
>    page.

Reviewer: S: there's a wiki role!
```suggestion
   To learn more about the ASCII standard, see the :wikipedia:`ASCII <w/index.php?title=ASCII&oldid=1180396712>` 
   Wikipedia page.
```

For example, in Canadian French, the right-most accented character determines
the ordering for strings when the other characters are the same. Consider the
following Canadian French words: 

- cote
- coté
- côte
- côté

When using the default binary collation, MongoDB sorts the words in the following order:

.. code-block:: none

   cote
   coté
   côte

>    côté

Reviewer: S: set copyable false for these two blocks
